BUG: handle multi-dimensional grouping better

dimithras opened the pull request:
As described in issue #36158 multi-dimensional groups prepared with lambda or named function raise error on indices as isnan() is not defined for MultiIndex.

Grouping already has the check for MultiIndex on line 440, but it's not triggered on __init__ when grouper is represented by **lambda** function and gets casted to MultiIndex later where MultiIndex stays as is.

~~Tested on the code provided in #36158, yet not sure if significant testing is needed as this is a relatively small change to code.  
Grouper code has not been changed for a while, should be save to merge.~~  
Tests for this particular issue provided.

~~**Concerns**: probably it will be better to make `isinstance(grouper, MultiIndex):` a function as after this change this piece of code occurs twice in __init__.~~  
Those operate on different grouper object. First one is passed on init, second one is `self`, where grouper is changed from lambda to MultiIndex and needs further change added in PR.

- [X] closes #36158
- [X] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

dimithras committed:
Resolve issue #36158

As described in issue #36158 multi-dimensional groups prepared with lambda raise error on indices as isnan() is not defined for MultiIndex.

Grouping already has the check for MultiIndex on line 440, but it's not triggered on __init__ when grouper is represented by lambda function and gets casted to MultiIndex later where MT stays as is.

arw2019 reviewed (changes requested):
thanks for the PR!

Could you add tests illustrating the new behavior? That's the first thing we look for when reviewing

dimithras committed:
Tests added for PR #36605

Bug in issue #36158 appeared both for indices and get_group().

dimithras committed:
Tests added for PR #36605

Bug in issue #36158 for indices and get_group().

dimithras commented:
Hi @arw2019 and thanks for initial review!
Updated PR, tests provided.

dimithras committed:
Aligned with [pre-commit] test

dimithras committed:
Added issue reference to tests

dimithras committed:
Aligned with [pre-commit] test

dimithras committed:
Comply with black

dimithras committed:
Missed `self.`

Missed `self.` during making PR, tests failed. Now corrected!

dimithras commented:
All tasks successful, tests passed, ***ready for review***.

jreback commented on pandas/core/groupby/grouper.py:
```diff
@@ -516,6 +516,10 @@ def __init__(
                     raise ValueError(f"Grouper for '{t}' not 1-dimensional")
                 self.grouper = self.index.map(self.grouper)
+
+                if isinstance(self.grouper, MultiIndex):
```
why do you think we should support a MultiIndex grouper at all?

dimithras replied:
1. Case for MultiIndex is already supported on line 440.
2. This bug is caused by grouper being identified as lambda. There might be other cases (line 515) where grouper will be re-identified and MultiIndex might be passed. In this case line 440 will be obsolete.

dimithras replied:
As for use cases - creating a multi-dimensional groupby, that is generally supported and listed in docs. [Please, refer to comment under issue](https://github.com/pandas-dev/pandas/issues/36158#issuecomment-696940968). Lambda, in particular, for cases when group columns are not included in a DataFrame and some custom logic is needed. Sure thing, columns with groups can be added to the original DataFrame, but lambda gives another approach which right now experiences a bug.

dimithras replied:
Last but not the least, grouping in general works well even with a bug present, only two functions are affected:
- indices
- get_group()

Also some other outputs described in diffs under [issue](https://github.com/pandas-dev/pandas/issues/36158#issuecomment-696940968) are affected as grouper uses MultiIndex where array is supposed to be present by current design (before modification in PR).

AlexanderCecile commented:
I thought I would mention this here, for visibility: The error also occurs with regular functions, not just `lambda`, see [my comment on the issue](https://github.com/pandas-dev/pandas/issues/36158#issuecomment-698595291).

dimithras commented:
@AlexanderCecile thank you very much for your input, highly appreciated. Fair point, behavior is the same both for lambda and defined function.  
I've just tested a function from [your comment](https://github.com/pandas-dev/pandas/issues/36158#issuecomment-697977693). Current solution also covers this case.

**Nature of a bug**:
With tuples in cells example using groupby by a column, grouper is initialized with actual values during grouper init ([line 413](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/grouper.py#L413)), this way MultiIndex is casted to an array on [line 440](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/grouper.py#L440).  
However if lambda or function is used, grouper is assigned lambda or named functioned respectively and [line 440](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/grouper.py#L440) is skipped as it does not see MultiIndex behind it.  
Later on [line 512](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/grouper.py#L512) changes are triggered and [line 518](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/grouper.py#L518) assigns actual value (MultiIndex) instead of a function, however this case skips [line 440](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/grouper.py#L440) casting and bug appears.

rhshadrach commented on pandas/core/groupby/grouper.py:
```diff
@@ -516,6 +516,10 @@ def __init__(
                     t = self.name or str(type(self.grouper))
                     raise ValueError(f"Grouper for '{t}' not 1-dimensional")
                 self.grouper = self.index.map(self.grouper)
```
Instead of using `map` it seems more appropriate to use `_transform_index`. Then no special handling of MultiIndex is required. The one groupby test that fails when doing this is `test_groupby_grouper_f_sanity_checked`, where the use of `map` here causes confusing errors down the line, which I _think_ is the reason for the length check on lines 523-526. By switching to `_transform_index`, the error message is instead `TypeError: 'Timestamp' object is not subscriptable` which looks to me to be the right error message.

dimithras replied:
Thanks @rhshadrach, this solves original issue in a more consistent way.

Yet, what should be done on a failing test?

rhshadrach replied:
If using `_transform_index` is correct here, then I think the test can be changed to check for a TypeError with the appropriate message to match. However, I realize now that `_transform_index` should only be used on callables - it's not clear to me from the code that `self.grouper` will always be a callable here.

dimithras replied:
@rhshadrach, if making a change to tests that do not relate to the issue under PR does not break any rules, I'm on it.
So far `_transform_index` seems to produce desired behavior, going to take a closer look.

rhshadrach replied:
Though it is not obvious from the issue being solve by this PR, the test is related to the fix - so it's okay to change. I would not remove the test, but rather make it pass by checking for the new (and sensible) error message.

dimithras committed:
Referencing changes in comment

Changes [from comment](https://github.com/pandas-dev/pandas/pull/36605#pullrequestreview-497010611) by @rhshadrach

dimithras committed:
Referencing changes in comment

Changes [from comment](https://github.com/pandas-dev/pandas/pull/36605#pullrequestreview-497010611) by @rhshadrach

rhshadrach commented:
@dimithras - Sorry, I missed your ping. If this is read for review, can you merge master and take this off draft status?

dimithras commented:
Hi @rhshadrach , sorry for letting this hanged. I looked at the failing test and issues connected to it, I'm more or less sure it should not be a problem to remove the test. Yet would be better to have a green light from you or someone else more familiar with pandas dev than me.

I'll remove the failing test tomorrow (CEST) so that it will be ready to merge.

dimithras committed:
Failing test edited

Previous assertion replaced with a new check.
Added description to linking issue.

Co-Authored-By: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

dimithras committed:
Failing test edited

Previous assertion replaced with a new check.
Added description to linking issue.

Co-Authored-By: Richard Shadrach <45562402+rhshadrach@users.noreply.github.com>

dimithras committed:
Merge branch 'patch-1' of https://github.com/dimithras/pandas into patch-1

dimithras commented:
Merged with master, no conflicts. Yet github tests are failing.

Tried running tests locally, I see many failures in `pandas/tests/arithmetic/test_datetime64.py` for instance. I see them both for master and the patch-1, could those be the reason?

@rhshadrach edited the grouping test, removing draft status for now.

dimithras commented:
Master merged with no conflicts, I also did a rebase to shrink the number of commits.  
@rhshadrach got it out of draft for now. Let me know if something needs to be done on Github's failing tests, do not see any clue why this is happening.

dimithras commented:
take

dimithras commented:
Converted to draft, **test_counting.py** `test_groupby_count_dateparseerror` is failing with same type error. Going to have a look at it.

dimithras committed:
Issue gone with master

After merging latest master issue is gone without applying the suggested fixes both by me and @rhshadrach .

Removing all changes yet leaving newly added tests referring to original issue.

dimithras commented:
After merging with latest master the issue is gone without making changes to grouper.

@rhshadrach unfortunately `_transform_index` breaks the test on test_counting.py in `test_groupby_count_dateparseerror` where lambda function is passed all multiindex items one by one instead of tuples. Instead of 10 tuples to 10 rows it gets 20 items where multiindex has two values.

All changes reverted, PR left with added tests for original issue.

dimithras committed:
Removed inconsistent use of pandas namespace

dimithras committed:
Update grouper.py

rhshadrach commented:
py37_32bit failed from out of memory error, Windows py38_np18 has one unrelated failure:
 
> FAILED pandas/tests/indexing/test_partial.py::TestPartialSetting::test_slice_irregular_datetime_index_with_nan

rhshadrach commented:
/azp run

rhshadrach reviewed (approved):
Thanks for updating, changes look good.

dimithras commented:
@jreback will you have a look if you have time?

jreback commented:
thanks @dimithras
